File: apollo-runtime/src/main/java/com/apollographql/apollo/response/ScalarTypeAdapters.java
package com.apollographql.apollo.response;

import com.apollographql.apollo.api.ScalarType;

import java.util.LinkedHashMap;
import java.util.Map;

import org.jetbrains.annotations.NotNull;

import static com.apollographql.apollo.api.internal.Utils.checkNotNull;

public final class ScalarTypeAdapters {
  private static final Map<Class, CustomTypeAdapter> DEFAULT_ADAPTERS = defaultAdapters();
  private final Map<ScalarType, CustomTypeAdapter> customAdapters;

  public ScalarTypeAdapters(@NotNull Map<ScalarType, CustomTypeAdapter> customAdapters) {
    this.customAdapters = checkNotNull(customAdapters, "customAdapters == null");
  }

  @SuppressWarnings("unchecked") @NotNull public <T> CustomTypeAdapter<T> adapterFor(@NotNull ScalarType scalarType) {
    checkNotNull(scalarType, "scalarType == null");

    CustomTypeAdapter<T> customTypeAdapter = customAdapters.get(scalarType);
    if (customTypeAdapter == null) {
      customTypeAdapter = DEFAULT_ADAPTERS.get(scalarType.javaType());
    }

    if (customTypeAdapter == null) {
      throw new IllegalArgumentException(String.format("Can't map GraphQL type: %s to: %s. Did you forget to add "
          + "custom type adapter?", scalarType.typeName(), scalarType.javaType()));
    }

    return customTypeAdapter;
  }

  private static Map<Class, CustomTypeAdapter> defaultAdapters() {
    Map<Class, CustomTypeAdapter> adapters = new LinkedHashMap<>();
    adapters.put(String.class, new DefaultCustomTypeAdapter<String>() {
      @NotNull @Override public String decode(@NotNull CustomTypeValue value) {
        return value.value.toString();
      }
    });
    adapters.put(Boolean.class, new DefaultCustomTypeAdapter<Boolean>() {
      @NotNull @Override public Boolean decode(@NotNull CustomTypeValue value) {
        if (value instanceof CustomTypeValue.GraphQLBoolean) {
          return (Boolean) value.value;
        } else if (value instanceof CustomTypeValue.GraphQLString) {
          return Boolean.parseBoolean(((CustomTypeValue.GraphQLString) value).value);
        } else {
          throw new IllegalArgumentException("Can't map: " + value + " to Boolean");
        }
      }
    });
    adapters.put(Integer.class, new DefaultCustomTypeAdapter<Integer>() {
      @NotNull @Override public Integer decode(@NotNull CustomTypeValue value) {
        if (value instanceof CustomTypeValue.GraphQLNumber) {
          return ((Number) value.value).intValue();
        } else if (value instanceof CustomTypeValue.GraphQLString) {
          return Integer.parseInt(((CustomTypeValue.GraphQLString) value).value);
        } else {
          throw new IllegalArgumentException("Can't map: " + value + " to Integer");
        }
      }
    });
    adapters.put(Long.class, new DefaultCustomTypeAdapter<Long>() {
      @NotNull @Override public Long decode(@NotNull CustomTypeValue value) {
        if (value instanceof CustomTypeValue.GraphQLNumber) {
          return ((Number) value.value).longValue();
        } else if (value instanceof CustomTypeValue.GraphQLString) {
          return Long.parseLong(((CustomTypeValue.GraphQLString) value).value);
        } else {
          throw new IllegalArgumentException("Can't map: " + value + " to Long");
        }
      }
    });
    adapters.put(Float.class, new DefaultCustomTypeAdapter<Float>() {
      @NotNull @Override public Float decode(@NotNull CustomTypeValue value) {
        if (value instanceof CustomTypeValue.GraphQLNumber) {
          return ((Number) value.value).floatValue();
        } else if (value instanceof CustomTypeValue.GraphQLString) {
          return Float.parseFloat(((CustomTypeValue.GraphQLString) value).value);
        } else {
          throw new IllegalArgumentException("Can't map: " + value + " to Float");
        }
      }
    });
    adapters.put(Double.class, new DefaultCustomTypeAdapter<Double>() {
      @NotNull @Override public Double decode(@NotNull CustomTypeValue value) {
        if (value instanceof CustomTypeValue.GraphQLNumber) {
          return ((Number) value.value).doubleValue();
        } else if (value instanceof CustomTypeValue.GraphQLString) {
          return Double.parseDouble(((CustomTypeValue.GraphQLString) value).value);
        } else {
          throw new IllegalArgumentException("Can't map: " + value + " to Double");
        }
      }
    });
    return adapters;
  }

  private abstract static class DefaultCustomTypeAdapter<T> implements CustomTypeAdapter<T> {
    @NotNull @Override public CustomTypeValue encode(@NotNull T value) {
      return CustomTypeValue.fromRawValue(value);
    }
  }
}

File: apollo-api-core/src/main/kotlin/com/apollographql/apollo/api/ScalarTypeAdapters.kt
package com.apollographql.apollo.api

class ScalarTypeAdapters(customAdapters: Map<ScalarType, CustomTypeAdapter<*>>) {
  private val customAdapters: MutableMap<String, CustomTypeAdapter<*>>

  init {
    this.customAdapters = HashMap(customAdapters.mapKeys { it.key.typeName() })
  }

  @Suppress("UNCHECKED_CAST")
  fun <T> adapterFor(scalarType: ScalarType): CustomTypeAdapter<T> {
    val customTypeAdapter: CustomTypeAdapter<*>? = (customAdapters[scalarType.typeName()] ?: DEFAULT_ADAPTERS[scalarType.javaType()])
    return requireNotNull(customTypeAdapter) {
      "Can't map GraphQL type: ${scalarType.typeName()} to: %${scalarType.javaType()}. Did you forget to add a custom type adapter?"
    } as CustomTypeAdapter<T>
  }

  private abstract class DefaultCustomTypeAdapter<T : Any> : CustomTypeAdapter<T> {

    override fun encode(value: T): CustomTypeValue<*> {
      return CustomTypeValue.fromRawValue(value)
    }
  }

  companion object {
    private val DEFAULT_ADAPTERS = LinkedHashMap<Class<*>, CustomTypeAdapter<*>>().apply {

      this[String::class.java] = object : DefaultCustomTypeAdapter<String>() {
        override fun decode(value: CustomTypeValue<*>): String {
          return value.value.toString()
        }
      }

      this[Boolean::class.java] = object : DefaultCustomTypeAdapter<Boolean>() {
        override fun decode(value: CustomTypeValue<*>): Boolean {
          return when (value) {
            is CustomTypeValue.GraphQLBoolean -> value.value
            is CustomTypeValue.GraphQLString -> value.value.toBoolean()
            else -> throw IllegalArgumentException("Can't decode: $value into Boolean")
          }
        }
      }

      this[Int::class.java] = object : DefaultCustomTypeAdapter<Int>() {
        override fun decode(value: CustomTypeValue<*>): Int {
          return when (value) {
            is CustomTypeValue.GraphQLNumber -> value.value.toInt()
            is CustomTypeValue.GraphQLString -> value.value.toInt()
            else -> throw IllegalArgumentException("Can't decode: $value into Integer")
          }
        }
      }

      this[Long::class.java] = object : DefaultCustomTypeAdapter<Long>() {
        override fun decode(value: CustomTypeValue<*>): Long {
          return when (value) {
            is CustomTypeValue.GraphQLNumber -> value.value.toLong()
            is CustomTypeValue.GraphQLString -> value.value.toLong()
            else -> throw IllegalArgumentException("Can't decode: $value into Long")
          }
        }
      }

      this[Float::class.java] = object : DefaultCustomTypeAdapter<Float>() {
        override fun decode(value: CustomTypeValue<*>): Float {
          return when (value) {
            is CustomTypeValue.GraphQLNumber -> value.value.toFloat()
            is CustomTypeValue.GraphQLString -> value.value.toFloat()
            else -> throw IllegalArgumentException("Can't decode: $value into Float")
          }
        }
      }

      this[Double::class.java] = object : DefaultCustomTypeAdapter<Double>() {
        override fun decode(value: CustomTypeValue<*>): Double {
          return when (value) {
            is CustomTypeValue.GraphQLNumber -> value.value.toDouble()
            is CustomTypeValue.GraphQLString -> value.value.toDouble()
            else -> throw IllegalArgumentException("Can't decode: $value into Double")
          }
        }
      }

      this[FileUpload::class.java] = object : CustomTypeAdapter<FileUpload> {
        override fun decode(value: CustomTypeValue<*>): FileUpload {
          throw UnsupportedOperationException("Can't decode file custom scalar type")
        }

        override fun encode(value: FileUpload): CustomTypeValue<*> {
          return CustomTypeValue.GraphQLString("")
        }
      }

      this[Any::class.java] = object : DefaultCustomTypeAdapter<Any>() {
        override fun decode(value: CustomTypeValue<*>): Any {
          return value.value!!
        }
      }

      this[java.lang.Object::class.java] = object : DefaultCustomTypeAdapter<Any>() {
        override fun decode(value: CustomTypeValue<*>): Any {
          return value.value!!
        }
      }

      this[Map::class.java] = object : DefaultCustomTypeAdapter<Map<*, *>>() {
        override fun decode(value: CustomTypeValue<*>): Map<*, *> {
          return if (value is CustomTypeValue.GraphQLJsonObject) {
            value.value
          } else {
            throw IllegalArgumentException("Can't decode: $value into Map")
          }
        }
      }

      this[List::class.java] = object : DefaultCustomTypeAdapter<List<*>>() {
        override fun decode(value: CustomTypeValue<*>): List<*> {
          return if (value is CustomTypeValue.GraphQLJsonList) {
            value.value
          } else {
            throw IllegalArgumentException("Can't decode: $value into List")
          }
        }
      }
    }
  }
}
